### PR TITLE
Stoch updates

### LINKED ISSUE #371: ocean stochastic physics
## Description
 Coupled ensemble data assimilation will need stochastic schemes to address model uncertainty in all components of the coupled system.  I first cut of ocean stochastic physics (SPPT and TKE perturbations in the Energetic PBL) have been developed and can get merged into the UFS.

## Solution
Additions of calls to ocean stochastic physics from MOM6 and modifications to code in parameterizations/vertical are necessary.
In addition, the capability of writing out stochastic physics restarts in netCDF at the restart interval is supported

## Related to
This issue will require PRs of
ccpp-physics, fv3atm, MOM6, and stochastic_physics
- required to support noaa-emc/nems/issues/<issue_number>
- needed by noaa-emc/fv3atm/pull/<pr_number>

### PR DESCRIPTION
**Pull request by pjpegion**
## Description
This PR references issue https://github.com/ufs-community/ufs-weather-model/issues/371

Addition of stochastic physics restarts in netCDF format that will be written at the restart interval, and at end of run.
Addition of stochastic cloud perturbation and Micropyiscs perturbations for SPPT.
and
clean up of stochastic physics code.

Is a change of answers expected from this PR?
No changes are expected with regression tests except the stochastic physics related regression tests. Stochastic physics regression test answer does change due to using a local version of ffpack instead of the fftpack in splib.

### Issue(s) addressed
- fixes https://github.com/ufs-community/ufs-weather-model/issues/371

## Testing
testing of new science additions were tested with 1-degree coupled model ensemble forecasts out to 35-day
What compilers / HPCs was it tested with? All tests on hera with intel and GNU compuler.
Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
The cloud perturbations and microphhysics perturbations are not included in any RTs.

Have the ufs-weather-model regression test been run? Currently a subset of regression tests were run on hera with intel and GNU compliers

Will the code updates change regression test baseline? If yes, why?
Yes, stochastic physics test will be changed due to using a local version of fftpack
I am waiting on code review before doing the full regression test suite.

- [x] hera.intel
- [x] hera.gnu
- [x] orion.intel
- [x] cheyenne.intel 
- [x] cheyenne.gnu
- [x] gaea.intel 
- [x] jet.intel
- [ ] wcoss_cray
- [ ] wcoss_dell_p3

new baseline date tag is 20210318.

## Dependencies

If testing this branch requires non-default branches in other repositories, list them.
Those branches should have matching names (ideally)

- waiting on https://github.com/NOAA-EMC/fv3atm/pull/233
- waiting on https://github.com/NCAR/ccpp-physics/pull/553
- waiting on https://github.com/noaa-psd/stochastic_physics/pull/35

### CODE REVIEW

File: tests/parm/input.mom6.nml.IN
Comment (DeniseWorthen) [inline tests/parm/input.mom6.nml.IN]: This file is used by ufs-weather when run with the datm. Is the addition of nam_stochy and sfcperts required here?

File: tests/rt.sh
Comment (DeniseWorthen) [inline tests/rt.sh]: We will need this set back to default

File: tests/default_vars.sh
Comment (DeniseWorthen) [inline tests/default_vars.sh]: My preference is to have no MOM6 flags or settings included in this commit rather than commenting them out.  Please either remove the optional flags you've added from the default_vars, edit_inputs and MOM_input templates or leave them present but ensure they are completely set false (or off).

Comment (jiandewang) [reply tests/default_vars.sh]: I agree with @DeniseWorthen, at least these setting shouldn't be presented in default_vars.sh

File: modulefiles/macosx.gnu/fv3_debug
Comment (junwang-noaa) [inline modulefiles/macosx.gnu/fv3_debug:42]: Was this PR tested on the maxosx platform?

File: tests/RegressionTests_hera.intel.log
Comment (junwang-noaa) [inline tests/RegressionTests_hera.intel.log]: Please resolve the conflict on the RT log files.

File: modulefiles/macosx.gnu/fv3_debug
Comment (climbfuji) [reply modulefiles/macosx.gnu/fv3_debug:42]: Not yet, but it will work, because the use of MKL has already been removed in a previous PR. mac is now also using the dcrft/scrft subroutines from libsp (? I think, or another nceplibs package).

File: MOM6-interface/CMakeLists.txt
Comment (DeniseWorthen) [inline MOM6-interface/CMakeLists.txt]: Are these CMakeList changes necessary if we are not committing the MOM6 stochastic options at this time?

File: CMakeLists.txt
Comment (climbfuji) [inline CMakeLists.txt:142]: Are all these changes required in this file? Seems like it involves a lot of changes for FMS. Is that because we are changing to a new version of FMS that has its own cmake build?

File: CMakeLists.txt
Comment (DusanJovic-NOAA) [inline CMakeLists.txt]: Remove commented if/endif

File: MOM6-interface/CMakeLists.txt
Comment (DusanJovic-NOAA) [inline MOM6-interface/CMakeLists.txt]: Remove

File: CMakeLists.txt
Comment (pjpegion) [reply CMakeLists.txt:142]: I pulled in https://github.com/ufs-community/ufs-weather-model/pull/441, so this may be part of that.

File: MOM6-interface/CMakeLists.txt
Comment (pjpegion) [reply MOM6-interface/CMakeLists.txt]: They are not necessary, but I had it working for when I thought the MOM6 stochastic physics PR was going in with this.

File: MOM6-interface/CMakeLists.txt
Comment (DusanJovic-NOAA) [inline MOM6-interface/CMakeLists.txt]: If `mom6solo` now depends on `stochastic_physics` target (line 87), is it necessary to explicitly add `../stochastic_physics` as an include directory. It shouldn't

**Comment by aerorahul** [comment]
@pjpegion 

@DusanJovic-NOAA asked me to review the cmake changes associated with this PR.  So here are my first impressions.

There are a number of unsavoury changes in the linking of stochastic physics library in the different parts of the code.
All of this is stemming from the fact that the stochastic_physics library created itself is not providing the appropriate and correct targets.
The changes in the `MOM6-interface/CMakeLists.txt` where the `target_include_directories` are pointing to `../stochastic_physics` etc would be unnecessary if we fixed the `CMakeLists.txt` in `stochastic_physics` itself.  I think that is where we should start.

After that, an option such as `USE_STOCHPHY_IN_MOM6` (just and e.g.) could be implemented or turned ON when the Ocean components are ready to turn that on and start linking to it.

I agree with the change in the way we build FMS.  FMS has a cmake build but we keep ignoring that and implement our own way. 

I am not sure we need to link `stochastic_physics` to the `ufs` library.  Is there an explicit dependency on it?  To my unfamiliar eyes, the dependency is in fv3atm and mom6.

File: MOM6-interface/CMakeLists.txt
Comment (DusanJovic-NOAA) [inline MOM6-interface/CMakeLists.txt]: Remove. `mom6` now depends on `stochastic_physics` target (line 67).

File: MOM6-interface/CMakeLists.txt
Comment (DusanJovic-NOAA) [inline MOM6-interface/CMakeLists.txt]: Remove.

File: MOM6-interface/CMakeLists.txt
Comment (DusanJovic-NOAA) [inline MOM6-interface/CMakeLists.txt]: Remove.

File: tests/parm/input.mom6.nml.IN
Comment (pjpegion) [reply tests/parm/input.mom6.nml.IN]: Not in the PR, but will be when the ocean stochastic physics is merged in.

**Review by DeniseWorthen — APPROVED** [review]
All my comments have been resolved. It needs updating to current develop but will approve in advance.

**Review by junwang-noaa — APPROVED** [review]

**Review by climbfuji — COMMENTED** [review]
Looks good, will approve once all the submodules are updated.

**Comment by MinsukJi-NOAA** [comment]
Should we merge #480 in this PR? Or in #471? CI will fail until #480 is merged in.

**Comment by DusanJovic-NOAA** [comment]
> Should we merge #480 in this PR? Or in #471? CI will fail until #480 is merged in.

I think we can merge #480 to this PR. @MinsukJi-NOAA please create PR to this branch. @pjpegion please merge Minsuk's PR. It will not impact the baseline creation we started.

File: tests/UnitTests_hera.intel.log
Comment (DusanJovic-NOAA) [inline tests/UnitTests_hera.intel.log]: Please resolve this conflict.

Comment (pjpegion) [reply tests/UnitTests_hera.intel.log]: This will be fixed when I upload the final regression test logs.

**Comment by DusanJovic-NOAA** [comment]
I do not see cheyenne log files.

**Comment by pjpegion** [comment]
> I do not see cheyenne log files.

I just pushed them.

**Comment by BrianCurtis-NOAA** [comment]
Machine: hera
Compiler: intel
Repo location: /scratch1/NCEPDEV/nems/emc.nemspara/autort/pr/555022768/20210319164512/ufs-weather-model
Please manually delete: /scratch1/NCEPDEV/stmp2/emc.nemspara/FV3_RT/rt_251895
Test fv3_gfsv16_ugwpv1_debug 075 failed in run_test failed 
Test fv3_gfdlmprad_noahmp 031 failed in run_test failed 
Test fv3_regional_quilt_hafs 027 failed in run_test failed 
Test fv3_regional_quilt 026 failed in run_test failed 
Test fv3_regional_quilt_netcdf_parallel 028 failed in run_test failed 
Test fv3_gfs_v15p2_debug 061 failed in run_test failed 
Test fv3_gfdlmprad_atmwav 018 failed in run_test failed 
Test fv3_gfs_v16_debug 062 failed in run_test failed 
Test fv3_gfdlmprad 017 failed in run_test failed 
Test cpld_controlfrac 079 failed in run_test failed 
Test cpld_control 077 failed in run_test failed 
Test cpld_2threads 081 failed in run_test failed 
Test fv3_gsd_debug 068 failed in run_test failed 
Test fv3_gfs_v15p2_RRTMGP_debug 063 failed in run_test failed 
Test cpld_decomp 082 failed in run_test failed 
Test cpld_satmedmf 083 failed in run_test failed 
Test cpld_control_c192 085 failed in run_test failed 
Test cpld_controlfrac_c192 087 failed in run_test failed 
Test cpld_ca 084 failed in run_test failed 
Test fv3_gfdlmprad_gwd 030 failed in run_test failed 
Test fv3_gfs_v16_restart 046 failed in run_test failed 
Test cpld_control_c384 089 failed in run_test failed 
Test cpld_bmark 093 failed in run_test failed 
Test cpld_controlfrac_c384 091 failed in run_test failed 
Test cpld_bmarkfrac 095 failed in run_test failed 
Test cpld_bmarkfrac_v16 097 failed in run_test failed 
Test datm_control_cfsr 105 failed in run_test failed 
Test datm_control_gefs 107 failed in run_test failed 
Test fv3_regional_restart 025 failed in run_test failed 
Test fv3_gfs_v16_RRTMGP_debug 064 failed in run_test failed 
Test datm_bulk_cfsr 108 failed in run_test failed 
Test datm_bulk_gefs 109 failed in run_test failed 
Test datm_mx025_cfsr 110 failed in run_test failed 
Test datm_mx025_gefs 111 failed in run_test failed 
Test fv3_stretched 022 failed in run_test failed 
Test fv3_control_debug 066 failed in run_test failed 
Test fv3_stretched_nest 023 failed in run_test failed 
Test fv3_gsd_diag3d_debug 069 failed in run_test failed 
Test fv3_regional_control_debug 065 failed in run_test failed 
Test fv3_thompson_no_aero_debug 071 failed in run_test failed 
Test fv3_thompson_debug 070 failed in run_test failed 
Test fv3_rrfs_v1beta_debug 072 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson_debug 073 failed in run_test failed 
Test fv3_gfs_v16_ras_debug 076 failed in run_test failed 
Test fv3_esg_HAFS_v0_hwrf_thompson_debug 074 failed in run_test failed 
Test fv3_stretched_nest_debug 067 failed in run_test failed 
Test fv3_gfs_v16_RRTMGP_c192L127 050 failed in run_test failed 
Test fv3_gocart_clm 054 failed in run_test failed 
Test fv3_wrtGauss_nemsio_c768 019 failed in run_test failed 
Test fv3_gfsv16_ugwpv1_warmstart 059 failed in run_test failed 
Test cpld_debug 103 failed in run_test failed 
Test cpld_debugfrac 104 failed in run_test failed 
Test datm_debug_cfsr 112 failed in run_test failed 
Test cpld_control_wave 102 failed in run_test failed 
Test fv3_gfs_v16_flake 055 failed in run_test failed 
Test cpld_bmarkfrac_wave 100 failed in run_test failed 
Test cpld_bmark_wave 099 failed in run_test failed 
Test cpld_bmarkfrac_wave_v16 101 failed in run_test failed 
Test fv3_HAFS_v0_hwrf_thompson 056 failed in run_test failed 
Test fv3_gfsv16_ugwpv1 058 failed in run_test failed 
Test fv3_gfs_v16_ras 060 failed in run_test failed 
Test fv3_esg_HAFS_v0_hwrf_thompson 057 failed in run_test failed 
Please make changes and add the following label back:
hera-intel-RT

**Comment by pjpegion** [comment]
@BrianCurtis-NOAA can you provide more infomation on what you are doing? I ran the hera.gnu test this morning, and I see that just up ran to too, and now I see that your hera.intel test failed, but mine worked.

**Comment by climbfuji** [comment]
> @BrianCurtis-NOAA can you provide more infomation on what you are doing? I ran the hera.gnu test this morning, and I see that just up ran to too, and now I see that your hera.intel test failed, but mine worked.

Duplication of efforts it seems ... hera just had intel compiler (license?) issues.

**Comment by DusanJovic-NOAA** [comment]
> @BrianCurtis-NOAA can you provide more infomation on what you are doing? I ran the hera.gnu test this morning, and I see that just up ran to too, and now I see that your hera.intel test failed, but mine worked.

It's auto-rt jobs that I submitted. It's not Brian.

**Comment by pjpegion** [comment]
@DusanJovic-NOAA thanks for the clarification

**Comment by BrianCurtis-NOAA** [comment]
Machine: jet
Compiler: intel
Repo location: /lfs4/HFIP/h-nems/emc.nemspara/autort/pr/555022768/20210319183026/ufs-weather-model
Please manually delete: /lfs4/HFIP/hfv3gfs/emc.nemspara/RT_RUNDIRS/emc.nemspara/FV3_RT/rt_287340
Test fv3_HAFS_v0_hwrf_thompson_debug 065 failed in run_test failed 
Test fv3_gfs_v16_ras_debug 068 failed in run_test failed 
Test fv3_gfsv16_ugwpv1_debug 067 failed in run_test failed 
Please make changes and add the following label back:
jet-intel-RT

**Comment by pjpegion** [comment]
Those three failed tests on jets have the same system error:

sbatch: error: Batch job submission failed: Socket timed out on send/recv operation


I compared the output of fv3_HAFS_v0_hwrf_thompson_debug and it has zero difference from baseline at FHR24.

**Comment by DusanJovic-NOAA** [comment]
> Those three failed tests on jets have the same system error:
> 
> sbatch: error: Batch job submission failed: Socket timed out on send/recv operation
> 
> I compared the output of fv3_HAFS_v0_hwrf_thompson_debug and it has zero difference from baseline at FHR24.

I reran the test manually. Log file is here:
/lfs4/HFIP/hfv3gfs/Dusan.Jovic/ufs/pr372/ufs-weather-model/tests/RegressionTests_jet.intel.log

**Pull request merged**
